networkd/network_server.ml

```diff
@@ -1012,13 +1013,31 @@ module PVS_proxy = struct
     open S.PVS_proxy
 
     let path = ref "/opt/citrix/pvsproxy/socket/pvsproxy"
-
-	let do_call call =
-		try
-			Jsonrpc_client.with_rpc ~path:!path ~call ()
-		with e ->
-			error "Error when calling PVS proxy: %s" (Printexc.to_string e);
-			raise PVS_proxy_connection_error
+    
+    let do_call call =
+            let timebox ~timeout f =
+                let fd_in, fd_out = Unix.pipe () in
+                let _ =  Thread.create
+                    (fun () ->
+                    f;
+                    debug "Complete execution in thread.";
+                    Unix.close fd_out) () in
+                if Thread.wait_timed_read fd_in timeout then
+                    debug "Timed execution completion."
+                else
+                    debug "Timed execution timed out.";
+                Unix.close fd_in
+            in
+            
+            let rpc_call =
```

Comment:
I don't understand... from what this code looks like, `rpc_call` is the actual value of the call, it's not running in a separate thread. Also these are OCaml threads, they are not really parallel. Also it seems that nothing is writing in `fd_in`, so that would stay blocked forever. Finally, I don't see a `Thread.join` or some other mean to make sure we execute and complete the thread. I think this code is not doing what it should

Reply:
Even if one adds the necessary `()` to run the call in the thread, I think this it is tricky to get right and the code above is not what we need. There is an infamous stackoverflow question that shows some options with different threadoffs and how tricky they are: https://stackoverflow.com/questions/35001959/interrupt-a-call-in-ocaml

Note in addition that one has to be very careful when dealing with `Thread.kill` and it generally not recommended.

Reply:
Would be better to add the timeout on the socket itself, you can try `setsockopt` to set timeouts on individual reads, or use `select` to wait for available input or a timeout (although thats not so easy to do because the code deals with buffered channels, not file descriptors directly)

Reply:
@mseri , this timebox call is following the style of implementation from the same named function timebox in ocaml/xapi/helpers.ml, which seems to be already used in xenopsd. And I did test both sunny day and rainy day scenario, refer to the ca ticket log, it works as expected. The issue you mentioned in the post seems not apply to the current implementation.

Reply:
@edwintorok , Rob suggest me to do this on the call, I guess the reason is Jon is trying to add something at lower level for the RPC call, like on socket as you mentioned.

Reply:
This is the `timebox` code from xapi. I doubt that the code above can work because the `Thread.create` does not apply function `f`. It is applied in the code below.

```
let timebox ~timeout ~otherwise f =
  let fd_in, fd_out = Unix.pipe () in
  let result = ref otherwise in
  let _ =  Thread.create
      (fun () ->
         (try
            let r = f () in
            result := fun () -> r
          with e ->
            result := fun () -> raise e);
         Unix.close fd_out) () in
  let _ = Thread.wait_timed_read fd_in timeout in
  Unix.close fd_in;
  !result ()
```

Reply:
@edwintorok I'm not sure exactly how `Thread.wait_timed_read` (from the standard library) is implemented, but it may just use `select` internally. The docs say "Suspend the execution of the calling thread until at least one character is available for reading ". This xapi function confuses me as well, because nothing is written to `fd_in`.

Reply:
Ah, here it is:
```
let wait_timed_read fd d =
  match Unix.select [fd] [] [] d with ([], _, _) -> false | (_, _, _) -> true
```

Reply:
I think closing the other side of the pipe wakes up the select, because there is now you can successfully read 0 bytes without blocking, i.e. EOF.

Reply:
https://github.com/ocaml/ocaml/pull/1656 lets improve the confusing documentation on `wait_timed_read`.